### ISSUE TITLE
[EFCore, Npgsql] ValueGeneratedOnAdd being incorrectly generated by Migration?

### ISSUE THREAD

**Issue by JakenVeina** [dotnet/efcore#16628]
The designer seems to be generating an incorrect ModelSnapshot for me, which results in an incorrect migration.

When generating a new migration, with no changes at all to source, I end up with this:

```cs
protected override void Up(MigrationBuilder migrationBuilder)
{
    migrationBuilder.AlterColumn<long>(
        name: "Id",
        table: "Users",
        nullable: false,
        oldClrType: typeof(long))
        .OldAnnotation("Npgsql:ValueGenerationStrategy", NpgsqlValueGenerationStrategy.SerialColumn);
}

protected override void Down(MigrationBuilder migrationBuilder)
{
    migrationBuilder.AlterColumn<long>(
        name: "Id",
        table: "Users",
        nullable: false,
        oldClrType: typeof(long))
        .Annotation("Npgsql:ValueGenerationStrategy", NpgsqlValueGenerationStrategy.SerialColumn);
}
```

Running this produces the following exception, due to the fact that the proposed sequence doesn't exist.

```
Applying migration '20190716015043_Initial'.
Applying migration '20190717003909_test'.
Failed executing DbCommand (15ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
ALTER TABLE "Users" ALTER COLUMN "Id" TYPE bigint;
ALTER TABLE "Users" ALTER COLUMN "Id" SET NOT NULL;
DROP SEQUENCE "Users_Id_seq" CASCADE;
ALTER TABLE "Users" ALTER COLUMN "Id" DROP DEFAULT;
Npgsql.PostgresException (0x80004005): 42P01: sequence "Users_Id_seq" does not exist
   at Npgsql.NpgsqlConnector.<>c__DisplayClass161_0.<<ReadMessage>g__ReadMessageLong|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Npgsql.NpgsqlConnector.<>c__DisplayClass161_0.<<ReadMessage>g__ReadMessageLong|0>d.MoveNext() in C:\projects\npgsql\src\Npgsql\NpgsqlConnector.cs:line 1032
--- End of stack trace from previous location where exception was thrown ---
   at Npgsql.NpgsqlDataReader.NextResult(Boolean async, Boolean isConsuming) in C:\projects\npgsql\src\Npgsql\NpgsqlDataReader.cs:line 490
   at Npgsql.NpgsqlDataReader.NextResult() in C:\projects\npgsql\src\Npgsql\NpgsqlDataReader.cs:line 332
   at Npgsql.NpgsqlCommand.ExecuteDbDataReader(CommandBehavior behavior, Boolean async, CancellationToken cancellationToken) in C:\projects\npgsql\src\Npgsql\NpgsqlCommand.cs:line 1218
   at Npgsql.NpgsqlCommand.ExecuteNonQuery(Boolean async, CancellationToken cancellationToken) in C:\projects\npgsql\src\Npgsql\NpgsqlCommand.cs:line 1042
   at Npgsql.NpgsqlCommand.ExecuteNonQuery() in C:\projects\npgsql\src\Npgsql\NpgsqlCommand.cs:line 1025
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, DbCommandMethod executeMethod, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteNonQuery(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationCommandExecutor.ExecuteNonQuery(IEnumerable`1 migrationCommands, IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.UpdateDatabase(String targetMigration, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
42P01: sequence "Users_Id_seq" does not exist
```

Looking at the `Initial` migration for those table, I can see that no generation is specified.

```cs
migrationBuilder.CreateTable(
    name: "Users",
    columns: table => new
    {
        Id = table.Column<long>(nullable: false),
        Username = table.Column<string>(nullable: false),
        Discriminator = table.Column<string>(nullable: false),
        AvatarHash = table.Column<string>(nullable: true),
        FirstSeen = table.Column<DateTimeOffset>(nullable: false),
        LastSeen = table.Column<DateTimeOffset>(nullable: false)
    },
    constraints: table =>
    {
        table.PrimaryKey("PK_Users", x => x.Id);
    });
```

However, in the `ModelSnapshot` both before and after adding the `test` migration, I see that a generator is specified.

```cs
modelBuilder.Entity("Sokan.Yastah.Data.Users.UserEntity", b =>
    {
        b.Property<long>("Id")
            .ValueGeneratedOnAdd();

        b.Property<string>("AvatarHash");

        b.Property<string>("Discriminator")
            .IsRequired();

        b.Property<DateTimeOffset>("FirstSeen");

        b.Property<DateTimeOffset>("LastSeen");

        b.Property<string>("Username")
            .IsRequired();

        b.HasKey("Id");

        b.ToTable("Users");
    });
```

The full definition for `UserEntity` can be found in the repository below.

Deleting and re-creating all migrations produces the same result.

[Full reproduction repository](https://github.com/JakenVeina/Sokan.Yastah/tree/efcore-issue-16628/Sokan.Yastah.Data)

Microsoft.EntityFrameworkCore v2.2.6
Microsoft.EntityFrameworkCore.Design v2.2.6
Npgsql.EntityFrameworkCore.PostgreSQL v2.2.4
Microsoft.NETCore.App v2.2.0
Visual Studio 2019 Community 16.1.1

**Comment by JakenVeina** [comment]
Also, I just noticed that there are two other entity types for which a value generator is being added, despite it not being in the model. `PermissionCategoryEntity` and `PermissionEntity`. These are presenting slightly differently, though, in that the `Initial` migration is self-consistent (it specifies that value generation is added). For these tables, running the migration doesn't result in an error, because a sequence is actually created during the `Initial` migration. It's just that a sequence shouldn't have been generated.

**Comment by ajcvickers** [comment]
@roji Does this look like it could be an Npgsql provider issue? Or more likely a core issue?

**Comment by ajcvickers** [comment]
Ping @roji

**Comment by roji** [comment]
Sorry for taking so long to get around to this.

Looking at the [repro repo](https://github.com/JakenVeina/Sokan.Yastah/tree/efcore-issue-16628/Sokan.Yastah.Data), the project seems to have switched to target .NET Core 3.0 (the issue was originally opened against 2.2.6), and uses the new support for PostgreSQL identity columns instead of serial. Adding a migration after cloning the repo generates an empty migrations as expected, and not the migration shown above. It's definitely possible that an issue in Npgsql caused a sync problem with the serial column sequence that could cause the error. Serial columns are complicated in that way, and that's precisely one of the main reasons PostgreSQL introduced the newer identity columns.

@JakenVeina can you please confirm whether you're still seeing any issues? In any case, if there's an issue it's very likely to be Npgsql-specific and so would belong in https://github.com/npgsql/efcore.pg.

**Comment by roji** [comment]
Looked into this again, and it seems to be a dupe of #11597: UserEntity has a value conversion defined for its Id (from ulong to long), but that scenario isn't currently supported. AFAIK this isn't supported to work with newer versions of EF Core or Npgsql, so it may be a good idea to check exactly what's happening in the latest versions of your application.

Am going to close this as a duplicate for now, but feel free to post back with more details - we can continue the conversation.

**Comment by JakenVeina** [comment]
This definitely shouldn't count as a duplicate of #11597 because while I am using a custom conversion on that column, I'ms specifically NOT using value-generation. And yet, the annotation was still being added, and more interestingly, the designer was generating a bogus migration.

Even if I made no changes at all to any model classes, adding a new migration would still generate content in the `Up()` and `Down()` methods, and trying to apply this migration would fail because it attempted to create objects that already exist.

Right now, everything seems to be resolved except for one thing: If I remove the `[DatabaseGenerated(DatabaseGeneratedOption.None)]` tag from `UserEntity.Id` (leaving only the `[Key]` tag) the designer still produces a `ModelSnapshot` that has `ValueGeneratedOnAdd()` for that column. If I do this for any other entity, this does not happen. Also, even though the `ModelSnapshot` includes a change in this case, the migration itself has empty `Up()` and `Down()` methods. So it seems the migration builder correctly interprets that there should not be a value generator, but the snapshot builder does not?

At the end of the day, it's a very easy workaround to just add the annotation, but it still definitely seems like a small bug to me.

**Comment by ajcvickers** [comment]
@JakenVeina Integer types are configured to use value generation by default. This means `UserEntity.Id` is configured to use value generation until you turn this off with `[DatabaseGenerated(DatabaseGeneratedOption.None)]`.

**Comment by JakenVeina** [comment]
Which was my first thought when revisiting this yesterday. Which is why I tried removing the `[DatabaseGenerated]` tag from some other entities that have generation disabled, so see if the designer added it back. And I could swear that it didn't, which is kinda the only reason I brought this back up.

I tried again just now to verify, and removing the tag on these columns does cause value generation to be enabled. So, I guess I'm just crazy.

Long story short, yeah, it looks like this was definitely resolved by 3.0. Sorry for the confusion.
